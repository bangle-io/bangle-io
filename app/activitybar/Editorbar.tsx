import React from 'react';

import {
  ActionButton,
  ButtonContent,
  TooltipWrapper,
} from '@bangle.io/ui-bangle-button';
import { CloseIcon, SecondaryEditorIcon } from '@bangle.io/ui-components';
import { cx } from '@bangle.io/utils';
import { removeExtension, resolvePath } from '@bangle.io/ws-path';

const MAX_ENTRIES = 3;

export function Editorbar({
  isActive,
  isSplitEditorOpen,
  onClose,
  onPressSecondaryEditor,
  openNotesPalette,
  showSplitEditor = false,
  wsPath,
}: {
  isActive: boolean;
  isSplitEditorOpen: boolean;
  onClose: () => void;
  onPressSecondaryEditor: () => void;
  openNotesPalette: () => void;
  showSplitEditor?: boolean;
  wsPath: string;
}) {
  let path = removeExtension(resolvePath(wsPath).filePath);

  if (path.split('/').length > MAX_ENTRIES) {
    let p = path.split('/').slice(-1 * MAX_ENTRIES);
    p.unshift('…');
    path = p.join('/');
  }

  if (path.length > 50) {
    path = '…' + path.slice(-1 * 50);
  }

  return (
    <div
      className="flex flex-row justify-between w-full B-activitybar_editorbar-wrapper px-2 py-1 lg:px-4 "
      style={{
        backgroundColor: 'var(--BV-window-bg-color-0)',
      }}
    >
      <div
        aria-label="note path"
        className={cx(
          'select-none flex flex-row flex-wrap text-xs cursor-pointer transition-colors px-2 rounded B-activitybar_editorbar-ws-path lg:text-sm text-ellipsis hover:underline',
          isActive && 'BU_active',
        )}
        onClick={openNotesPalette}
      >
        {path}
      </div>
      <div className="flex flex-row flex-1"></div>
      <div className="flex flex-row">
        {showSplitEditor && (
          <ActionButton
            isQuiet="hoverBg"
            onPress={onPressSecondaryEditor}
            className="lg:mr-1"
            ariaLabel="Split screen"
            isActive={isSplitEditorOpen}
            tooltip={
              <TooltipWrapper>
                {isSplitEditorOpen ? 'Close split screen' : 'Split screen'}
              </TooltipWrapper>
            }
            tooltipPlacement="bottom"
            tooltipXOffset={10}
          >
            <ButtonContent size="small" icon={<SecondaryEditorIcon />} />
          </ActionButton>
        )}
        <ActionButton
          isQuiet="hoverBg"
          ariaLabel="Close"
          onPress={onClose}
          styling={{}}
        >
          <ButtonContent size="small" icon={<CloseIcon />} />
        </ActionButton>
      </div>
    </div>
  );
}
