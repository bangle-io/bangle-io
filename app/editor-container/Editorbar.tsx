import React from 'react';

import { TONE } from '@bangle.io/constants';
import { vars } from '@bangle.io/css-vars';
import type { WsPath } from '@bangle.io/shared-types';
import {
  Button,
  BUTTON_VARIANT,
  CloseIcon,
  SecondaryEditorIcon,
} from '@bangle.io/ui-components';
import { cx } from '@bangle.io/utils';
import { removeExtension, resolvePath2 } from '@bangle.io/ws-path';

const MAX_ENTRIES = 3;

export function Editorbar({
  isActive,
  isSplitEditorOpen,
  onClose,
  onEnableEditing,
  onPressSecondaryEditor,
  openNotesPalette,
  showSplitEditor = false,
  wsPath,
  editingDisabled,
}: {
  isActive: boolean;
  isSplitEditorOpen: boolean;
  onClose: () => void;
  onPressSecondaryEditor: () => void;
  openNotesPalette: () => void;
  onEnableEditing: () => void;
  editingDisabled?: boolean;
  showSplitEditor?: boolean;
  wsPath: WsPath;
}) {
  let path = removeExtension(resolvePath2(wsPath).filePath);

  if (path.split('/').length > MAX_ENTRIES) {
    let p = path.split('/').slice(-1 * MAX_ENTRIES);
    p.unshift('…');
    path = p.join('/');
  }

  if (path.length > 50) {
    path = '…' + path.slice(-1 * 50);
  }

  return (
    <div
      data-testid="app-editor-container_editorbar"
      className="flex flex-row justify-between w-full px-2 py-1 lg:px-4"
      style={{
        backgroundColor: vars.misc.editorBg,
      }}
    >
      <Button
        ariaLabel="note path"
        text={path}
        size="xs"
        className={isActive ? 'BU_active' : ''}
        tone={TONE.SECONDARY}
        variant={!isActive ? BUTTON_VARIANT.TRANSPARENT : BUTTON_VARIANT.SOFT}
        onPress={openNotesPalette}
      />

      {editingDisabled && (
        <Button
          ariaLabel="enable editing"
          text="Enable Editing"
          className="mx-2"
          size="xs"
          tone={TONE.PROMOTE}
          onPress={onEnableEditing}
        />
      )}
      <div className="flex flex-row flex-1"></div>
      <div className="flex flex-row">
        {showSplitEditor && (
          <Button
            size="xs"
            variant={isSplitEditorOpen ? 'soft' : 'transparent'}
            onPress={onPressSecondaryEditor}
            className={cx('lg:mr-1', isSplitEditorOpen ? 'BU_is-active' : '')}
            ariaLabel={
              isSplitEditorOpen ? 'Close split screen' : 'Split screen'
            }
            leftIcon={<SecondaryEditorIcon />}
            tooltipPlacement="bottom"
          />
        )}
        <Button
          size="xs"
          variant="transparent"
          ariaLabel="Close"
          onPress={onClose}
          leftIcon={<CloseIcon />}
        />
      </div>
    </div>
  );
}
