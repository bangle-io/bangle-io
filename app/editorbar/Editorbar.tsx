import React from 'react';

import {
  ActionButton,
  ButtonContent,
  TooltipWrapper,
} from '@bangle.io/ui-bangle-button';
import { CloseIcon, SecondaryEditorIcon } from '@bangle.io/ui-components';
import { cx } from '@bangle.io/utils';
import { removeExtension, resolvePath } from '@bangle.io/ws-path';

const MAX_ENTRIES = 3;

export function Editorbar({
  showSplitEditor = false,
  wsPath,
  onClose,
  onPressSecondaryEditor,
  isSplitEditorOpen,
  isActive,
  openNotesPalette,
}: {
  isActive: boolean;
  showSplitEditor?: boolean;
  wsPath: string;
  onClose: () => void;
  onPressSecondaryEditor: () => void;
  isSplitEditorOpen: boolean;
  openNotesPalette: () => void;
}) {
  let path = removeExtension(resolvePath(wsPath).filePath);

  let p = path.split('/');

  if (p.length > MAX_ENTRIES) {
    p = p.slice(-1 * MAX_ENTRIES);
    p.unshift('â€¦');
  }

  return (
    <div className="flex flex-row justify-between w-full B-editorbar_wrapper ">
      <div
        aria-label="note path"
        className={cx(
          'flex flex-row flex-wrap text-xs cursor-pointer transition-colors px-2 rounded B-editorbar_ws-path lg:text-sm text-ellipsis hover:underline',
          isActive && 'BU_active',
        )}
        onClick={openNotesPalette}
      >
        {p.map((r, i) => (
          <React.Fragment key={i}>
            <span className="break-all select-none">{r}</span>
            {i !== p.length - 1 && (
              <span className="select-none" style={{ padding: '0 1px' }}>
                /
              </span>
            )}
          </React.Fragment>
        ))}
      </div>
      <div className="flex flex-row">
        {showSplitEditor && (
          <ActionButton
            isQuiet="hoverBg"
            onPress={onPressSecondaryEditor}
            styling={{}}
            className="lg:mr-1"
            ariaLabel="Split screen"
            isActive={isSplitEditorOpen}
            tooltip={
              <TooltipWrapper>
                {isSplitEditorOpen ? 'Close split screen' : 'Split screen'}
              </TooltipWrapper>
            }
            tooltipPlacement="bottom"
            tooltipXOffset={10}
          >
            <ButtonContent size="small" icon={<SecondaryEditorIcon />} />
          </ActionButton>
        )}
        <ActionButton
          isQuiet="hoverBg"
          ariaLabel="Close"
          onPress={onClose}
          styling={{}}
        >
          <ButtonContent size="small" icon={<CloseIcon />} />
        </ActionButton>
      </div>
    </div>
  );
}
