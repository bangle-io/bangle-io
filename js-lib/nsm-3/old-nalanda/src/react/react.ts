import { useDebugValue, useRef, useState } from 'react';
import useSyncExternalStoreExports from 'use-sync-external-store/shim';

import type { Slice } from '../vanilla/slice';
import type { Store } from '../vanilla/store';
import type { AnySlice } from '../vanilla/types';

const { useSyncExternalStore } = useSyncExternalStoreExports;

type InferData<T> = T extends Slice<any, infer TData, any> ? TData : never;

interface ReactAdapter<TSnapshot> {
  subscribe: (onStoreChange: () => void) => () => void;
  getSnapshot: () => TSnapshot;
}

export function createUseTrackSliceHook<TAllSliceName extends string = any>(
  store: Store<TAllSliceName>,
) {
  function useSlice<TSlice extends AnySlice>(sl: TSlice): InferData<TSlice> {
    const ref = useRef<InferData<TSlice>>();

    const [adapter] = useState(() => {
      const adapter: ReactAdapter<InferData<TSlice>> = {
        subscribe: (onStoreChange) => {
          const sliceEffect = store.effect((effectStore) => {
            const selectedData = sl.track(effectStore);
            ref.current = selectedData;

            onStoreChange();

            queueMicrotask(() => {
              if (effectStore._runInstance?._addTrackedCount === 0) {
                console.warn(
                  `You forgot to destructure/access the tracked field from ${sl.name}. This will not track any changes!.`,
                );
              }
            });
          });

          return () => {
            store.unregisterEffect(sliceEffect);
          };
        },
        getSnapshot: () => {
          return ref.current ?? sl.get(store.state);
        },
      };

      return adapter;
    });

    const data = useSyncExternalStore(adapter.subscribe, adapter.getSnapshot);

    useDebugValue(data);

    return data;
  }

  return useSlice;
}
