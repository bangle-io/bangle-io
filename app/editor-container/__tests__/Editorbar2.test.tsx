/**
 * @jest-environment @bangle.io/jsdom-env
 */
import { fireEvent, render } from '@testing-library/react';
import React from 'react';

import { Editorbar } from '../Editorbar';

test('renders correctly', () => {
  const openNotesPalette = jest.fn();
  let result = render(
    <div>
      <Editorbar
        openNotesPalette={openNotesPalette}
        isActive={false}
        showSplitEditor={false}
        wsPath={'mojo:test-dir/magic.md'}
        onClose={jest.fn()}
        onPressSecondaryEditor={jest.fn()}
        isSplitEditorOpen={false}
        onEnableEditing={jest.fn()}
      />
    </div>,
  );

  expect(result.getByLabelText('note path').className).not.toContain(
    'BU_active',
  );

  expect(result.container).toMatchSnapshot();
});

test('renders correctly when active', () => {
  let result = render(
    <div>
      <Editorbar
        openNotesPalette={jest.fn()}
        isActive={true}
        showSplitEditor={false}
        wsPath={'mojo:test-dir/magic.md'}
        onClose={jest.fn()}
        onPressSecondaryEditor={jest.fn()}
        isSplitEditorOpen={false}
        onEnableEditing={jest.fn()}
      />
    </div>,
  );

  expect(result.container).toMatchSnapshot();
});

test('truncates large wsPath', () => {
  let result = render(
    <div>
      <Editorbar
        openNotesPalette={jest.fn()}
        isActive={false}
        showSplitEditor={false}
        wsPath={'mojo:test-dir/magic/wow/last/two.md'}
        onClose={jest.fn()}
        onPressSecondaryEditor={jest.fn()}
        isSplitEditorOpen={false}
        onEnableEditing={jest.fn()}
      />
    </div>,
  );

  expect(result.getByLabelText('note path')).toMatchSnapshot();
});

test('dispatches togglePaletteType on clicking wsPath', () => {
  const openNotesPalette = jest.fn();
  let result = render(
    <div>
      <Editorbar
        openNotesPalette={openNotesPalette}
        isActive={false}
        showSplitEditor={true}
        wsPath={'mojo:test-dir/magic.md'}
        onClose={jest.fn()}
        onPressSecondaryEditor={jest.fn()}
        isSplitEditorOpen={true}
        onEnableEditing={jest.fn()}
      />
    </div>,
  );

  fireEvent.click(result.getByLabelText('note path'));

  expect(openNotesPalette).toBeCalledTimes(1);
});

test('renders splitscreen', () => {
  let result = render(
    <div>
      <Editorbar
        openNotesPalette={jest.fn()}
        isActive={false}
        showSplitEditor={true}
        wsPath={'mojo:test-dir/magic.md'}
        onClose={jest.fn()}
        onPressSecondaryEditor={jest.fn()}
        isSplitEditorOpen={true}
        onEnableEditing={jest.fn()}
      />
    </div>,
  );

  expect(result.getByLabelText('Close split screen')).toBeDefined();
  result.rerender(
    <div>
      <Editorbar
        openNotesPalette={jest.fn()}
        isActive={false}
        showSplitEditor={true}
        wsPath={'mojo:test-dir/magic.md'}
        onClose={jest.fn()}
        onPressSecondaryEditor={jest.fn()}
        isSplitEditorOpen={false}
        onEnableEditing={jest.fn()}
      />
    </div>,
  );
  expect(result.getByLabelText('Split screen')).toBeDefined();
});

test('enables editing', () => {
  const onEnableEditing = jest.fn();
  let result = render(
    <div>
      <Editorbar
        openNotesPalette={jest.fn()}
        isActive={false}
        showSplitEditor={true}
        wsPath={'mojo:test-dir/magic.md'}
        onClose={jest.fn()}
        onPressSecondaryEditor={jest.fn()}
        isSplitEditorOpen={true}
        onEnableEditing={onEnableEditing}
        editingDisabled
      />
    </div>,
  );

  fireEvent.click(result.getByLabelText('enable editing'));

  expect(onEnableEditing).toBeCalledTimes(1);
});
