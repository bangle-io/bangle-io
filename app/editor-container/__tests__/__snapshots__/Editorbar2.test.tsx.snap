// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div>
    <div
      class="flex flex-row justify-between w-full px-2 py-1 lg:px-4"
      data-testid="app-editor-container_editorbar"
    >
      <button
        aria-label="note path"
        class="text-xs font-600 min-w-6 h-6 px-1_5  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
        style="background-color: transparent;"
        type="button"
      >
        <span
          class="flex flex-grow-1 overflow-hidden "
          style="justify-content: center;"
        >
          <span
            class="text-ellipsis overflow-hidden "
          >
            test-dir/magic
          </span>
        </span>
      </button>
      <div
        class="flex flex-row flex-1"
      />
      <div
        class="flex flex-row"
      >
        <button
          aria-label="Close"
          class="text-xs font-600 min-w-6 h-6 px-1_5  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
          style="background-color: transparent;"
          type="button"
        >
          <span
            class="flex flex-grow-1 overflow-hidden "
            style="justify-content: center;"
          >
            <span>
              <svg
                class=" w-4 h-4  "
                fill="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6.2253 4.81108C5.83477 4.42056 5.20161 4.42056 4.81108 4.81108C4.42056 5.20161 4.42056 5.83477 4.81108 6.2253L10.5858 12L4.81114 17.7747C4.42062 18.1652 4.42062 18.7984 4.81114 19.1889C5.20167 19.5794 5.83483 19.5794 6.22535 19.1889L12 13.4142L17.7747 19.1889C18.1652 19.5794 18.7984 19.5794 19.1889 19.1889C19.5794 18.7984 19.5794 18.1652 19.1889 17.7747L13.4142 12L19.189 6.2253C19.5795 5.83477 19.5795 5.20161 19.189 4.81108C18.7985 4.42056 18.1653 4.42056 17.7748 4.81108L12 10.5858L6.2253 4.81108Z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly when active 1`] = `
<div>
  <div>
    <div
      class="flex flex-row justify-between w-full px-2 py-1 lg:px-4"
      data-testid="app-editor-container_editorbar"
    >
      <button
        aria-label="note path"
        class="BU_active text-xs font-600 min-w-6 h-6 px-1_5  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
        type="button"
      >
        <span
          class="flex flex-grow-1 overflow-hidden "
          style="justify-content: center;"
        >
          <span
            class="text-ellipsis overflow-hidden "
          >
            test-dir/magic
          </span>
        </span>
      </button>
      <div
        class="flex flex-row flex-1"
      />
      <div
        class="flex flex-row"
      >
        <button
          aria-label="Close"
          class="text-xs font-600 min-w-6 h-6 px-1_5  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
          style="background-color: transparent;"
          type="button"
        >
          <span
            class="flex flex-grow-1 overflow-hidden "
            style="justify-content: center;"
          >
            <span>
              <svg
                class=" w-4 h-4  "
                fill="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6.2253 4.81108C5.83477 4.42056 5.20161 4.42056 4.81108 4.81108C4.42056 5.20161 4.42056 5.83477 4.81108 6.2253L10.5858 12L4.81114 17.7747C4.42062 18.1652 4.42062 18.7984 4.81114 19.1889C5.20167 19.5794 5.83483 19.5794 6.22535 19.1889L12 13.4142L17.7747 19.1889C18.1652 19.5794 18.7984 19.5794 19.1889 19.1889C19.5794 18.7984 19.5794 18.1652 19.1889 17.7747L13.4142 12L19.189 6.2253C19.5795 5.83477 19.5795 5.20161 19.189 4.81108C18.7985 4.42056 18.1653 4.42056 17.7748 4.81108L12 10.5858L6.2253 4.81108Z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`truncates large wsPath 1`] = `
<button
  aria-label="note path"
  class="text-xs font-600 min-w-6 h-6 px-1_5  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
  style="background-color: transparent;"
  type="button"
>
  <span
    class="flex flex-grow-1 overflow-hidden "
    style="justify-content: center;"
  >
    <span
      class="text-ellipsis overflow-hidden "
    >
      â€¦/wow/last/two
    </span>
  </span>
</button>
`;
